ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+
+.. image:: /images/colocation.png
+  :alt: colocation in citus
+
+In order to populate ``http_request_1min`` we're going to periodically run the equivalent
+of an INSERT INTO SELECT. Citus doesn't yet support INSERT INTO SELECT on distributed
+tables, so instead we'll run a function on all the workers which runs INSERT INTO SELECT
+on every matching pair of shards (this is possible because the shards are colocated)
+
+.. code-block:: plpgsql
+
+  -- this should be run on each worker
+  CREATE FUNCTION rollup_1min(source_shard text, dest_shard text) RETURNS void
+  AS $$
+  DECLARE
+    v_latest_minute_already_aggregated timestamptz;
+    v_new_latest_already_aggregated timestamptz;
+  BEGIN
+    PERFORM SET lock_timeout 100;
+    -- since master calls this function every minute, and future invokations will
+    -- do any work this function doesn't do, it's safe to quit if we wait too long
+    -- for this FOR UPDATE lock which makes sure at most one instance of this function
+    -- runs at a time
+    SELECT ingest_time INTO v_latest_minute_already_aggregated FROM rollup_thresholds
+      WHERE source_shard = source_shard AND dest_shard = dest_shard
+      FOR UPDATE;
+    IF NOT FOUND THEN
+      INSERT INTO rollup_thresholds VALUES (
+        '1970-01-01', source_shard::regclass, dest_shard::regclass);
+      RETURN;
+    END IF;
+    PERFORM RESET lock_timeout;
+
+    EXECUTE format('
+      WITH (
+        INSERT INTO %I (
+            zone_id, ingest_time, request_count,
+            error_count, success_count, average_response_time_msec)
+          SELECT
+            zone_id,
+            date_trunc('minute', ingest_time) as minute,
+            COUNT(1) as request_count,
+            COUNT(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
+            request_count - success_count AS error_count,
+            SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
+            -- later sections will show some more clauses that can go here
+          FROM %I
+          WHERE ingest_time > v_latest_minute_already_aggregated
+          GROUP BY zone_id, minute
+        ) as inserted_rows
+      SELECT max(minute) INTO v_new_latest_already_aggregated FROM inserted_rows;
+    ', dest_shard, source_shard);
+
+    -- mark how much work we did, so the next invocation picks up where we left off
+    PERFORM UPDATE rollup_thresholds
+      SET ingest_time = v_new_latest_already_aggregated
+      WHERE source_shard = source_shard AND dest_shard = dest_shard;
+  END;
+  $$ LANGUAGE 'plpgsql';
+
+As discussed above, there's a 1-to-1 correspondence between http_request shards and
+``http_request_1min`` shards. This function accepts the name of the ``http_request`` shard
+to read from and the name of the ``http_request_1min`` shard to write to. It can't figure
+it out itself because that kind of metadata is kept on the master, not the workers.
```

Comment:
Lots of its. Maybe omit the sentence, or "The worker doesn't know which shard to use, since the Citus metadata is only kept on the master."